`noUnusedLocals`/`noUnusedParameters`

(Thanks for sharing this lint rule! ❤️)

**Repro**

Reduced test case: https://github.com/OliverJAsh/eslint-type-tests-test

1. Run `yarn`
2. Run `eslint index.ts`

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "noUnusedLocals": false,
    "noUnusedParameters": false
  }
}
```

`foo.ts`:
```ts
export const foo = 1;
```

`index.ts`:
```ts
import { foo } from './foo';

// $ExpectType 1
type F = typeof foo;
```

**Expected Result**

No lint error

**Actual Result**

```sh-session
$ eslint index.ts

/Users/oliverash/Development/eslint-type-tests-test/index.ts
  53:2  error  TypeScript compile error: 'F' is declared but never used  expect-type/expect

✖ 1 problem (1 error, 0 warnings)
```

**Additional Info**

It looks like this happens because `typescript-estree` enables `noUnusedLocals`/ `noUnusedParameters`. Code here: https://github.com/typescript-eslint/typescript-eslint/blob/bf904ec72db57174fec531f61e9427230662553e/packages/typescript-estree/src/create-program/shared.ts#L19-L23

It looks like there is no way to disable these compiler options, even if they are specified as `false` in the TS config (as I have done in my reduced test case).

In this case I don't want ESLint to error if something is unused. It's sometimes helpful to declare a type/variable which is "unused", so we can hover over the declaration to inspect its type.

Perhaps this lint rule should filter out these TS errors if `noUnusedLocals`/`noUnusedParameters` are not enabled by the local project?

--- a/src/rules/expect.ts
+++ b/src/rules/expect.ts
@@ -228,7 +228,18 @@ function validate(context: TSESLint.RuleContext<MessageIds, [Options]>, options:
     });
   }
 
+  function diagnosticShouldBeIgnored(diagnostic: ts.Diagnostic) {
+    const messageText = typeof diagnostic.messageText === "string"
+      ? diagnostic.messageText
+      : diagnostic.messageText.messageText
+    return /'.+' is declared but (never used|its value is never read)./.test(messageText);
+  }
+
   function addDiagnosticFailure(diagnostic: ts.Diagnostic): void {
+    if (diagnosticShouldBeIgnored(diagnostic)) {
+      return;
+    }
+
     if (diagnostic.file === sourceFile) {
       const message = `${ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n')}`;
       context.report({

--- a/tests/rules/expect-type.test.ts
+++ b/tests/rules/expect-type.test.ts
@@ -25,6 +25,16 @@ const valid: ReadonlyArray<ValidTestCase> = [
     `,
     optionsSet: [[]],
   },
+  // Ignored TypeScript compiler complaints
+  {
+    code: dedent`
+      function hasUnusedParam(unusedParam: number, implicitAnyParam) {
+        const unusedLocal = 0;
+        return implicitAnyParam;
+      }
+    `,
+    optionsSet: [[]],
+  }
 ];
 
 // Invalid test cases.

--- a/tests/sandbox/tsconfig.json
+++ b/tests/sandbox/tsconfig.json
@@ -3,6 +3,12 @@
     "lib": ["es2017"],
     "module": "commonjs",
     "moduleResolution": "node",
+
+    // Testing that these don't get passed through as issues
+    // https://github.com/ibezkrovnyi/eslint-plugin-expect-type/issues/16
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+
     "skipLibCheck": true,
     "target": "es2017"
   },

